compiler/damlc/daml-stdlib-src/DA/Internal/Template.daml

```diff
@@ -6,60 +6,42 @@
 
 daml 1.2
 -- | MOVE Prelude DAML-LF primitives, just templates/contracts
-module DA.Internal.Template(
-    Template(ensure, signatory, observer, agreement),
-    Choice(consuming, choiceController, choice), preconsuming, nonconsuming, postconsuming, NoEvent(..),
-    stakeholder,
-    create, exercise, fetch,
-    archive, Archive(..),
-    lookupByKey, fetchByKey, exerciseByKey,
-    IsParties(toParties),
-    TemplateKey(key, maintainer)
-    ) where
-
-import GHC.Types (magic)
+module DA.Internal.Template where
+
 import DA.Internal.LF
 import DA.Internal.Prelude
 
--- NOTE: use internalCreate vs create so that people implementing Template can't override create/fetch/exercise
-
--- | Create a contract based on a template.
-create : Template c => c -> Update (ContractId c)
-create = internalCreate
-
--- | Exercise a contract choice.
-exercise : forall c e r . Choice c e r => ContractId c -> e -> Update r
-exercise = internalExercise
-
--- | Fetch the contract data associated with the given contract id.
--- This fails and aborts the entire transaction if the `ContractId c`
--- supplied is not the contract ID of an active contract.
-fetch : Template c => ContractId c -> Update c
-fetch = internalFetch
+-- | The stakeholders of a contract, i.e. its signatories and observers.
+stakeholder : Template t => t -> [Party]
+stakeholder t = signatory t ++ observer t
 
--- | Archive the contract.
-archive : Template c => ContractId c -> Update ()
-archive c = exercise c Archive
+class Template t where
+  signatory : t -> [Party]
+  observer : t -> [Party]
+  ensure : t -> Bool
+  agreement : t -> Text
+  create : t -> Update (ContractId t)
+  fetch : ContractId t -> Update t
+  archive : ContractId t -> Update ()
 
--- | The stakeholders of a contract, i.e. its signatories and observers.
-stakeholder : Template c => c -> [Party]
-stakeholder c = signatory c ++ observer c
+class Template t => Choice t c r | t c -> r where
+  exercise : ContractId t -> c -> Update r
 
 -- | Look up the contract id associated with a given contract key.
 --
--- The `c` needs to be passed using an explicit type application. For
+-- The `t` needs to be passed using an explicit type application. For
 -- instance, if you want to look up a contract of template `Account` by its
 -- key `k`, you must call `lookupByKey @Account k`.
-lookupByKey : forall c k. TemplateKey c k => k -> Update (Optional (ContractId c))
-lookupByKey = internalLookupByKey
+-- lookupByKey : forall t k. TemplateKey t k => k -> Update (Optional (ContractId t))
```

Comment:
Why is this commented out? Same for `fetchByKey`.

Reply:
These are provided by the `TemplateKey` instances now, right? Is there any value in having separate function definitions?

Reply:
Right, good point. Let's remove them here.